File: backends/elixir-phoenix/test/cursor_test.exs
defmodule StackverseBackend.CursorTest do
  use ExUnit.Case, async: true

  alias StackverseBackend.Cursor

  test "cursor round-trips created_at and id" do
    created_at = DateTime.from_naive!(~N[2026-07-05 12:30:00], "Etc/UTC")
    id = "00000000-0000-4000-8000-000000000001"

    raw = Cursor.encode(%{created_at: created_at, id: id})

    assert {:ok, %{created_at: ^created_at, id: ^id}} = Cursor.decode(raw)
    assert :error = Cursor.decode("definitely-not-a-cursor")
  end
end

File: backends/ruby-rails/test/services/stackverse/cursor_test.rb
require "test_helper"

class StackverseCursorTest < ActiveSupport::TestCase
  test "round trips opaque bookmark cursors" do
    cursor = Stackverse::BookmarkCursor.new(Time.utc(2026, 7, 5, 12, 30, 15, 123456), "00000000-0000-0000-0000-000000000001")

    decoded = Stackverse::Cursor.decode(Stackverse::Cursor.encode(cursor))

    assert_equal cursor.id, decoded.id
    assert_equal cursor.created_at.iso8601(6), decoded.created_at.utc.iso8601(6)
  end

  test "rejects malformed cursors as bad requests" do
    error = assert_raises(Stackverse::ProblemError) { Stackverse::Cursor.decode("not a cursor") }

    assert_equal 400, error.problem.status
  end
end
